### ISSUE TITLE
Build and run QtCore tests

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#29]
I don't want us to repeat the Mozilla for OS/2 experience where its internal test framework is miles behind because no-one cared about it in years. So I will make the Qt test suite work right from the beginning here.

This might include some additional porting.

**Comment by dmik** [comment]
There are many tests based on the QtTest module, so apparently we need to build it first.

**Comment by dmik** [comment]
All Qt autotests based on QtTest live in `tests/auto` (the QtCore ones live in `corelib`, obviously). Technically, these tests (also including benchmark tests living in `tests/benchmarks`) are designed to be run *after* the whole `qtbase` module built. Since we only build parts of it, we can't go this way. And we don't want to postpone tests until we build all `qtbase` as it's better to reveal e.g. QtCore bugs earlier since many other things in Qt depend on it.

A solution here is a custom make file that looks similar to this one (SRCROOT and BUILDROOT speak for themselves I guess — only note that the former is a path to the parent directory containing the qt5 root repository, not the repository's dir itself):
```make
tests: CURDIR = $(BLDROOT)/qtbase
tests:
	@test -d $(CURDIR)/tests/ || mkdir -p $(CURDIR)/tests/
	cd $(CURDIR)/tests/ && ( test -e Makefile || $(QMAKE) -o Makefile $(SRCROOT)/qt5/qtbase/tests/tests.pro ) && $(MAKE) -C auto/corelib -f Makefile debug
```

Most corelib tests build, just a few ones fail to compile. Will sort them out.

**Comment by dmik** [comment]
These `corelib` tests still fail to build because of the incomplete `qtbase`:
- `tests/auto/corelib/animation/qpropertyanimation` (needs `Qt5Widgets[d].lib`)
- `tests/auto/corelib/kernel/qeventloop` (needs `Qt5Network[d].lib`)
- `tests/auto/corelib/io/qfile/test` (needs `Qt5Network[d].lib`)

There is also a number of warnings but they seem to be upstream and not relevant (some are even intentional). We will ignore them.

**Comment by dmik** [comment]
These `corelib` tests still fail to build because of the incomplete `qtbase` module:

- `tests/auto/corelib/animation/qpropertyanimation` (needs `Qt5Widgets[d].lib`)
- `tests/auto/corelib/kernel/qeventloop` (needs `Qt5Network[d].lib`)
- `tests/auto/corelib/io/qfile/test` (needs `Qt5Network[d].lib`)
- `tests/auto/corelib/io/qiodevice` (needs `Qt5Network[d].lib`)

There is also a number of warnings but they seem to be upstream and not relevant (some are even intentional). We will ignore them.

**Comment by dmik** [comment]
I don't like the idea of disabling tests that depend on other modules and running the others. Especially given that the key tests like `qfile` are affected. And even though the QtNetwork module dependency is really dumb (getting the host name for some UNC path testing IIRC), I don't want to touch test sources just because of that and then revert it back once QtNetwork is done.

So it was decided to move tests in a separate project. We will work on it as we go with Qt modules. I will get back to QtCore tests once I finish QtNetwork (which is my next task now (#35).

**Comment by dmik** [comment]
There is also a bunch of tests failing because of the lack of `Qt5Guild[d].lib`. I have to build it too it seems.

**Comment by dmik** [comment]
I found a nice way to avoid disabling individual tests in project files: just configure Qt with `-no-gui -no-widgets`. This effectively disables QtGui/QtWidgets and test .pro files are smart enough to figure that out and not build the respective tests.

I'm still building the corelib test suite though. There seems to be a zillion of them. I will commit my small build patches at once while I'm there.

**Comment by dmik** [comment]
Finally, all corelib tests are built. There are total of 224 executables. Now I will try to run them all and see which still fail.

**Comment by dmik** [comment]
Unfortunately, I can't tell how many tests fail: they are run by a Makefile and any unexpected failure breaks the bunch. Most commonly I see hangs now (Ctrl-C killable). I have to debug these first. And also I see `QEventDispatcherOS2Private: DosWaitThread failed 309` from some code that belongs to the test framework itself.

**Comment by dmik** [comment]
The `DosWaitThread` warning has been fixed (see the above commit). However, I also see some tests crash with the following:
```
PASS   : tst_QAbstractAnimation::avoidJumpAtStart()
terminate called after throwing an instance of '__gnu_cxx::recursive_init_error'
  what():  std::exception

Killed by SIGABRT
pid=0xb7cc ppid=0xb7cb tid=0x0003 slot=0x007f pri=0x0200 mc=0x0001 ps=0x0010
D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\ANIMATION\QABSTRACTANIMATION\DEBUG\TST_QABSTRACTANIMATION.EXE
```
It's not always reproducible but can be eventually seen after running the same test many times. This exception is a special one, see http://gcc.gnu.org/onlinedocs/gcc-4.6.0/libstdc%2B%2B/api/a00063.html#_details and some explanation here https://stackoverflow.com/questions/6967179/whats-a-recursive-init-error-exception. It looks that somehow we go recursive when initializing some static Qt object. I have to find this.

**Comment by dmik** [comment]
Okay, I guess I know what's going on. First of all, I started getting this exception after I added some `qDebug` statements to the code that starts an aux thread for socket/timer event processing. I managed to run the problem under debugger and spotted that the exception is thrown when two `qDebug` calls made from two different threads (the main one and the aux one) overlap in time. This overlap happens when both threads call `QLoggingCategory::defaultCategory()` nearly at the same time (which takes place not on each run, of course). This method involves some static object constructor magic (rolled out by Q_GLOBAL_STATIC) in order to initialize a singleton upon its first use. More over, the `QLoggingCategory` constructor instantiates another singleton via `QLoggingRegistry::instance()` and this is where it eventually goes down due to some recursion somewhere (it ends up with XCPT_BAD_STACK in the debugger if not termnated).

Qt assumes that all GCC compilers greater version 3.4 thread-safe static object initialization. However, it seems not to be the case in our OS/2 port of GCC 4.9.2. At least, if I make Qt use the other code path that is present there for non-thread-safe compilers (it involves a mutex) the problem goes away. I cannot reproduce it any more after a few dozen of runs, at least.

So I will mark that GCC on OS/2 is not thread-safe in this regard (this should make all tests as well as all real-life cases involving Q_GLOBAL_STATIC happy) and create a GCC ticket to further track this issue later.

**Comment by dmik** [comment]
The next problem is that some tests start but hang up too early w/o even running the first test function, like this:
```
make.exe[1]: Entering directory `D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/animation/qparallelanimationgroup'
D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/animation/qparallelanimationgroup/target_wrapper.sh  debug/tst_qparallelanimationgroup.exe
********* Start testing of tst_QParallelAnimationGroup *********
Config: Using QtTest library 5.11.0, Qt 5.11.0 (i386-little_endian-ilp32 shared (dynamic) debug build; by GCC 4.9.2)
```
Something bad happens somewhere at the very beginning of the test harness which doesn't let it start. Need to find out.

**Comment by dmik** [comment]
I've used the following script to run tests w/o make stopping on each failure:
``` sh
#! /bin/sh

export QT_MESSAGE_PATTERN='[%{time process}:%{threadid}] %{function}:%{line}: %{message}'
export QTEST_FUNCTION_TIMEOUT=30000

find . -name target_wrapper.sh -execdir make check \; > run-tests.log 2>&1
```
and the following script to get some statistics and filter out test results:
``` sh
#! /bin/sh

grep 'Totals: ' -c run-tests.log
grep -E '(([1-9])|([0-9]+[0-9])) failed,' -c run-tests.log
grep -A1 Totals: run-tests.log > run-tests-filter.log
```

The results are as following: 178 total tests and 36 have failures. I'm also attaching two logs here: the first one is with short test results (contains the name of the test and the number of failures) and the second one is the full test transcript (necessary to see which sub-tests failed, exactly).

[run-tests-filter.log](https://github.com/bitwiseworks/qtbase-os2/files/2452442/run-tests-filter.log)
[run-tests.log](https://github.com/bitwiseworks/qtbase-os2/files/2452443/run-tests.log)

**Comment by dmik** [comment]
Here is a slightly updated filter script:
``` sh
#! /bin/sh

grep 'Totals: ' -c run-tests.log
grep -E '(([1-9])|([0-9]+[0-9])) failed,' -c run-tests.log
grep -A1 -E '(([1-9])|([0-9]+[0-9])) (failed|skipped|blacklisted),' run-tests.log > run-tests-filter.log
```
and the contents of the filtered list inline (it's 36 failed tests + ones that don't have failures but have skipped and blacklisted subtests — these also need a look):
```
Totals: 14 passed, 1 failed, 0 skipped, 0 blacklisted, 1674ms
********* Finished testing of tst_QAbstractAnimation *********
--
Totals: 1 passed, 1 failed, 0 skipped, 0 blacklisted, 30037ms
********* Finished testing of tst_QPauseAnimation *********
--
Totals: 8 passed, 2 failed, 0 skipped, 0 blacklisted, 31026ms
********* Finished testing of tst_QSequentialAnimationGroup *********
--
Totals: 39 passed, 0 failed, 1 skipped, 0 blacklisted, 28ms
********* Finished testing of tst_QGetPutEnv *********
--
Totals: 57 passed, 1 failed, 1 skipped, 0 blacklisted, 33ms
********* Finished testing of tst_QNumeric *********
--
Totals: 45 passed, 11 failed, 0 skipped, 0 blacklisted, 61ms
********* Finished testing of tst_LargeFile *********
--
Totals: 320 passed, 34 failed, 7 skipped, 0 blacklisted, 514ms
********* Finished testing of tst_QDir *********
--
Totals: 17 passed, 8 failed, 0 skipped, 0 blacklisted, 625ms
********* Finished testing of tst_QDirIterator *********
--
Totals: 178 passed, 26 failed, 6 skipped, 0 blacklisted, 2383ms
********* Finished testing of tst_QFileInfo *********
--
Totals: 18 passed, 3 failed, 0 skipped, 0 blacklisted, 29ms
********* Finished testing of tst_QFileSelector *********
--
Totals: 62 passed, 6 failed, 0 skipped, 0 blacklisted, 16ms
********* Finished testing of tst_QFileSystemEntry *********
--
Totals: 15 passed, 9 failed, 0 skipped, 0 blacklisted, 56351ms
********* Finished testing of tst_QFileSystemWatcher *********
--
Totals: 69 passed, 0 failed, 2 skipped, 0 blacklisted, 601ms
********* Finished testing of tst_QIODevice *********
--
Totals: 9 passed, 9 failed, 5 skipped, 0 blacklisted, 2639ms
********* Finished testing of tst_QLockFile *********
--
Totals: 12 passed, 1 failed, 0 skipped, 0 blacklisted, 62ms
********* Finished testing of tst_QSaveFile *********
--
Totals: 5 passed, 4 failed, 0 skipped, 0 blacklisted, 24ms
********* Finished testing of tst_qstandardpaths *********
--
Totals: 3 passed, 8 failed, 0 skipped, 0 blacklisted, 27ms
********* Finished testing of tst_QStorageInfo *********
--
Totals: 28 passed, 2 failed, 0 skipped, 0 blacklisted, 1361ms
********* Finished testing of tst_QTemporaryDir *********
--
Totals: 50 passed, 1 failed, 0 skipped, 0 blacklisted, 484ms
********* Finished testing of tst_QTemporaryFile *********
--
Totals: 819 passed, 4 failed, 1 skipped, 0 blacklisted, 268ms
********* Finished testing of tst_QUrl *********
--
Totals: 18 passed, 1 failed, 0 skipped, 0 blacklisted, 31099ms
********* Finished testing of tst_QCoreApplication *********
--
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 31003ms
********* Finished testing of tst_QEventLoop *********
--
Totals: 28 passed, 2 failed, 0 skipped, 0 blacklisted, 30229ms
********* Finished testing of tst_QObject *********
--
Totals: 2 passed, 2 failed, 0 skipped, 0 blacklisted, 30064ms
********* Finished testing of tst_QSocketNotifier *********
--
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 30035ms
********* Finished testing of tst_QTimer *********
--
Totals: 65 passed, 0 failed, 5 skipped, 0 blacklisted, 521ms
********* Finished testing of tst_QMimeDatabase *********
--
Totals: 1 passed, 1 failed, 0 skipped, 0 blacklisted, 23ms
********* Finished testing of tst_QFactoryLoader *********
--
Totals: 17 passed, 14 failed, 0 skipped, 0 blacklisted, 43ms
********* Finished testing of tst_QLibrary *********
--
Totals: 6 passed, 9 failed, 0 skipped, 0 blacklisted, 64ms
********* Finished testing of tst_QPluginLoader *********
--
Totals: 857 passed, 3 failed, 1 skipped, 0 blacklisted, 23016ms
********* Finished testing of tst_QTextStream *********
--
Totals: 149 passed, 1648 failed, 0 skipped, 0 blacklisted, 2447ms
********* Finished testing of tst_QXmlStream *********
--
Totals: 7 passed, 1 failed, 0 skipped, 0 blacklisted, 32233ms
********* Finished testing of tst_QStateMachine *********
--
Totals: 34 passed, 1 failed, 0 skipped, 0 blacklisted, 31707ms
********* Finished testing of tst_QMutex *********
--
Totals: 8 passed, 1 failed, 0 skipped, 0 blacklisted, 35254ms
********* Finished testing of tst_QThread *********
--
Totals: 383 passed, 0 failed, 1 skipped, 0 blacklisted, 115ms
********* Finished testing of tst_QByteArray *********
--
Totals: 19651 passed, 0 failed, 1 skipped, 0 blacklisted, 1328ms
********* Finished testing of tst_QChar *********
--
Totals: 46 passed, 2 failed, 0 skipped, 0 blacklisted, 30121ms
********* Finished testing of tst_QCommandLineParser *********
--
Totals: 438 passed, 0 failed, 33 skipped, 0 blacklisted, 572ms
********* Finished testing of tst_QDateTime *********
--
Totals: 1 passed, 1 failed, 0 skipped, 0 blacklisted, 7ms
********* Finished testing of tst_QLocale *********
--
Totals: 51 passed, 1 failed, 0 skipped, 0 blacklisted, 33152ms
********* Finished testing of tst_QSharedPointer *********
--
Totals: 1359 passed, 0 failed, 14 skipped, 0 blacklisted, 590ms
********* Finished testing of tst_QString *********
--
Totals: 15 passed, 0 failed, 3 skipped, 0 blacklisted, 31ms
********* Finished testing of tst_QStringView *********
--
Totals: 10733 passed, 0 failed, 1 skipped, 0 blacklisted, 502ms
********* Finished testing of tst_QTextBoundaryFinder *********
--
Totals: 2 passed, 1 failed, 0 skipped, 0 blacklisted, 30139ms
********* Finished testing of tst_QTimeLine *********
--
Totals: 87 passed, 4 failed, 0 skipped, 0 blacklisted, 2192ms
********* Finished testing of tst_QTimeZone *********
```

**Comment by dmik** [comment]
Fixing kLIBC took quite a while (as always) but now there are only 9 failed sub-tests in the `tst_QDir` test and 340 succeed (yes, this test is huge, which is expected). The rest of failures should be easy to fix.

**Comment by dmik** [comment]
Note that I'm currently debugging the `QProcess` class as it shows some strange errors (can't start symlinked executables). The tst_QDir test case also revealed that.

**Comment by dmik** [comment]
With `QDir::canonicalPath` fixed, `tst_QDir` now finally completes with the following stats:
```
Totals: 343 passed, 0 failed, 5 skipped, 0 blacklisted, 1819ms
```
Not bad, I must say. The 5 skipped test cases are the ones creating file paths with file names of 256 chars and longer (which expectedly fail on many other platforms too so these failures are converted to skips by the test case).

**Comment by dmik** [comment]
Note that I fixed some mkspecs bug in 333166d which made `tst_QFile` completely fail and sneak out of the log excerpt in the comment above. I've just made it run and putting its stats to that comment. Quite a lot of failures, expected.

**Comment by dmik** [comment]
Another major test case, tst_QFileInfo, bit the dust.

**Comment by dmik** [comment]
BTW, the tst_QFileSystemMetaData test was also lost somehow (due to a complete failure I believe). But it now works so I added it to the list.

**Comment by dmik** [comment]
JFYI, all `io` tests in `corelib` are done and that was the most complex part I must say. Now working on `kernel` tests and struggling with some QObject issues.

**Comment by dmik** [comment]
Note that making timers work fixed all three failing animation test cases which is not a surprise as they use timers a lot.

**Comment by dmik** [comment]
Just for the record, `tst_QSocketNotifier::posixSockets` is not very accurate as it installs a signal spy on a read notification from a socket that checks that the signal was emitted only once. However, the test case is written so that the socket is not read in the socket activation signal (which is guaranteed to happen before the select thread starts selecting on it again). This, of course, assumes that the socket may be reported as ready for read twice — it depends on timings (i.e. whether the actual socket read happens before the select thread starts selecting on it again, or after). This means that the given test case fails on a comparison at #315 every 5 runs or so. As I consider to be the test case bug, I will not address it further.

**Comment by dmik** [comment]
Testing is almost finished, there are just a few failures in a few tests remaining to be fixed. `tst_QStateMachine/QMutex/QThread` have some thread related failures. `tst_QDateTime` seems to freak out on DST checks for years beyond 2038 (can be a limitation of our LIBC, some 32 bit field somewhere or such). `tst_QLocale` triggers a SEGFPU when doing some conversion to double, I guess this is related to #22. `tst_QSharedPointer` calls the compiler (!) and linker to build some code on the fly and this fails because of some non-resolved library paths (as we don't install Qt5 when testing it). `tst_QTimeLine` has some multithreaded issue too. And `tst_QTimeZone` has one DST problem again.

**Comment by SilvanScherrer** [comment]
Cant we just ignore the failure of qdatetime? As I doubt 2038 with qt5 on os2 will ever happen 🤣

**Comment by dmik** [comment]
Both QThread and QMutex tests work well now. For QMutex it turned out that it has an enormous run time (more than 200 sec) so `QTEST_FUNCTION_TIMEOUT` should be set pretty high (the default value of 300 sec should also be fine).

**Comment by dmik** [comment]
With #22 being fixed, `tst_QLocale` works flawlessly now. It still freaks out in one test when setting the central European time zone, but this needs a deeper investigation, as well as the other failing TZ cases. Created #58 for that. Besides this, only `tst_QStateMachine` and `tst_QSharedPointer` remain.

**Comment by dmik** [comment]
Okay, all but the TZ tests (#58) have been finished now. However, tst_QTimer::timerInfiniteRecursion broke due to some recent change, I suspect 8125510bcca9def12f4e8a2f6b0c3bbdaaa4ade8. Checking.

**Comment by dmik** [comment]
I was right above — it was a regression of 8125510. I fixed it but found another failed test case — tst_QSocketNotifier::asyncMultipleDatagram. Checking it too.

**Comment by dmik** [comment]
The socket and timer problems have gone now but I discovered two tests that somehow slipped out of my sight: `tst_QSystemSemaphore` and `tst_QSharedMemory`, both fail so far. I need to give them a quick look. As the corresponding entities are ported and should work.

**Comment by dmik** [comment]
`tst_QSystemSemaphore` works now, in `tst_QSharedMemory` there are still 3 failures (out of 51 passed tests). Will knock them down too (remember, these are all-new classes).

**Comment by dmik** [comment]
Note that *ALL* tests should run fine with the debug build of Qt now. It's time to check the release build.

**Comment by dmik** [comment]
Unfortunately, an attempt to run all tests at once at once gave a few failures:
```
FAIL!  : tst_QSequentialAnimationGroup::startGroupWithRunningChild() Compared values are not the same
   Actual   (anim1->state())          : Running
   Expected (QAnimationGroup::Stopped): Stopped
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/animation/qsequentialanimationgroup/tst_qsequentialanimationgroup.cpp(1301)]
FAIL!  : tst_qmessagehandler::setMessagePattern() Compared values are not the same
   Actual   (QString::fromLatin1(output))  : "static constructor\r\n[debug] qDebug\r\n[info] qInfo\r\n[warning] qWarning\r\n[critical] qCritical\r\n[warning] qDebug with category\r\n"
   Expected (QString::fromLatin1(expected)): "static constructor\n[debug] qDebug\n[info] qInfo\n[warning] qWarning\n[critical] qCritical\n[warning] qDebug with category\n"
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/global/qlogging/tst_qlogging.cpp(896)]
FAIL!  : tst_QProcess::startDetached() 'QProcess::startDetached("testProcessNormal/testProcessNormal", QStringList() << "arg1" << "arg2")' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(308)]
FAIL!  : tst_QProcess::detachedProcessParameters(none) 'process.startDetached(&pid)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2133)]
FAIL!  : tst_QProcess::detachedProcessParameters(stdout) 'process.startDetached(&pid)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2133)]
FAIL!  : tst_QProcess::detachedProcessParameters(stderr) 'channelFile.remove()' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2100)]
FAIL!  : tst_QTimer::basic_chrono() 'helper.count > 0' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/kernel/qtimer/tst_qtimer.cpp(254)]
FAIL!  : tst_QSemaphore::tryAcquireWithTimeout(0.2s) 'qAbs(a1-e1) < fuzz' returned FALSE. ((time.elapsed()=0) is more than 60 milliseconds different from (timeout=200))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/thread/qsemaphore/tst_qsemaphore.cpp(347)]
FAIL!  : tst_QSemaphore::tryAcquireWithTimeout(2s) 'qAbs(a1-e1) < fuzz' returned FALSE. ((time.elapsed()=1589) is more than 150 milliseconds different from (timeout=2000))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/thread/qsemaphore/tst_qsemaphore.cpp(326)]
FAIL!  : tst_QThread::wait2() 'elapsed >= Waiting_Thread::WaitTime - 10' returned FALSE. (elapsed: 0)
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/thread/qthread/tst_qthread.cpp(1086)]
FAIL!  : tst_QThread::wait2() Received a fatal error.
   Loc: [Unknown file(0)]
FAIL!  : tst_QThreadStorage::initTestCase() 'QFileInfo(m_crashOnExit).isExecutable()' returned FALSE. (D:\Coding\qt5\qt5-dev-build\qtbase\tests\auto\corelib\thread\qthreadstorage\crashonexit\crashonexit does not exist or is not executable.)
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/thread/qthreadstorage/tst_qthreadstorage.cpp(89)]
FAIL!  : tst_QTimeZone::isTimeZoneIdAvailable() 'QTimeZone::isTimeZoneIdAvailable(id)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(375)]
FAIL!  : tst_QTimeZone::specificTransition(Moscow/2014) Compared values are not the same
   Actual   (transits.length()): 0
   Expected (count)            : 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(455)]
FAIL!  : tst_QTimeZone::specificTransition(Moscow/2011) Compared values are not the same
   Actual   (transits.length()): 0
   Expected (count)            : 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(455)]
FAIL!  : tst_QTimeZone::transitionEachZone(America/Bahia_Banderas@2010) Received a fatal error.
   Loc: [Unknown file(0)]
```
This is really not much given the amount of 66738 (yes, 66 thousand 738) PASSED tests in 185 tests units.

I will check this.

**Comment by dmik** [comment]
A couple of tests have simply never been run before, fixed them above.

One test `tst_QMetaType` is not there as it crashed with SIGFPE and slipped out of the awk filter expression. I reopened https://github.com/bitwiseworks/libcx/issues/56 to address that.

`tst_QTimeZone` failures are to be addressed within #58.

`tst_QProcess` failures are really strange as I can't reproduce them when running `tst_QProcess` alone. I need a scenario and a debug version of LIBCx to see what's going on there.

As for the rest, these are purely timing failures, I created #67 for them. Looks like QTBUGs to me.

Note that this is still a debug build. I will now try to run a release build and see what it gives us. I expect that some issues from #67 will go away at least.

**Comment by dmik** [comment]
With the above commits, we have 68651 tests from 189 test suites succeed and only `tst_QProcess::startDetached`, `tst_QProcess::detachedProcessParameters`, `tst_QTemporaryFile::io` (also timings) plus `tst_QTimeZone` tests failing. Not bad, I must say.

**Comment by dmik** [comment]
The release build fails in a few additional tests, unfortunately:
```
FAIL!  : tst_QProcess::startDetached() 'QProcess::startDetached("testProcessNormal/testProcessNormal", QStringList() << "arg1" << "arg2")' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(308)]
FAIL!  : tst_QProcess::detachedProcessParameters(none) 'process.startDetached(&pid)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2133)]
FAIL!  : tst_QProcess::detachedProcessParameters(stdout) 'process.startDetached(&pid)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2133)]
FAIL!  : tst_QProcess::detachedProcessParameters(stderr) 'channelFile.remove()' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/io/qprocess/tst_qprocess.cpp(2100)]
FAIL!  : tst_QSharedMemory::initTestCase() '!m_helperBinary.isEmpty()' returned FALSE. (Could not find helper binary)
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/kernel/qsharedmemory/test/tst_qsharedmemory.cpp(139)]
FAIL!  : tst_QSystemSemaphore::initTestCase() '!m_helperBinary.isEmpty()' returned FALSE. (Could not find helper binary)
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/kernel/qsystemsemaphore/test/tst_qsystemsemaphore.cpp(81)]
FAIL!  : tst_QLibrary::load(ok00) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/mylib: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\MYLIB.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(215)]
FAIL!  : tst_QLibrary::load(ok01 (with suffix)) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/mylib.dll: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\MYLIB.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(215)]
FAIL!  : tst_QLibrary::load(ok02 (with non-standard suffix)) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/mylib.dl2: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\MYLIB.DL2.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(215)]
FAIL!  : tst_QLibrary::load(ok03 (with many dots)) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/qt.mylib.dll: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\QT.MYLIB.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(215)]
FAIL!  : tst_QLibrary::resolve(ok00) 'func != 0' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(281)]
FAIL!  : tst_QLibrary::unload(mylib) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/mylib: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\MYLIB.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(242)]
FAIL!  : tst_QLibrary::unload_after_implicit_load() 'p' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(252)]
FAIL!  : tst_QLibrary::errorString(bad resolve) Compared values are not the same
   Actual   (ok)  : 0
   Expected (true): 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(381)]
FAIL!  : tst_QLibrary::errorString(good resolve) Compared values are not the same
   Actual   (ok)  : 0
   Expected (true): 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(381)]
FAIL!  : tst_QLibrary::loadHints(ok01 (with suffix)) Compared values are not the same
   Actual   (library.loadHints()): <null>
   Expected (lh)                 : ResolveAllSymbolsHint
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(446)]
FAIL!  : tst_QLibrary::loadHints(ok02 (with non-standard suffix)) Compared values are not the same
   Actual   (library.loadHints()): <null>
   Expected (lh)                 : ResolveAllSymbolsHint
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(446)]
FAIL!  : tst_QLibrary::loadHints(ok03 (with many dots)) Compared values are not the same
   Actual   (library.loadHints()): <null>
   Expected (lh)                 : ResolveAllSymbolsHint
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(446)]
FAIL!  : tst_QLibrary::fileName(ok02) 'ok' returned FALSE. (Cannot load library D:/Coding/qt5/qt5-dev-build/qtbase/tests/auto/corelib/plugin/qlibrary/release/mylib.dll: (dlopen rc=2 extra=D:\CODING\QT5\QT5-DEV-BUILD\QTBASE\TESTS\AUTO\CORELIB\PLUGIN\QLIBRARY\RELEASE\MYLIB.DLL))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(482)]
FAIL!  : tst_QLibrary::multipleInstancesForOneLibrary() Compared values are not the same
   Actual   (lib1.isLoaded()): 0
   Expected (true)           : 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/plugin/qlibrary/tst_qlibrary.cpp(501)]
FAIL!  : tst_QSemaphore::tryAcquireWithTimeout(0.2s) 'qAbs(a1-e1) < fuzz' returned FALSE. ((time.elapsed()=285) is more than 60 milliseconds different from (timeout=200))
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/thread/qsemaphore/tst_qsemaphore.cpp(383)]
FAIL!  : tst_QTimeZone::isTimeZoneIdAvailable() 'QTimeZone::isTimeZoneIdAvailable(id)' returned FALSE. ()
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(375)]
FAIL!  : tst_QTimeZone::specificTransition(Moscow/2014) Compared values are not the same
   Actual   (transits.length()): 0
   Expected (count)            : 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(455)]
FAIL!  : tst_QTimeZone::specificTransition(Moscow/2011) Compared values are not the same
   Actual   (transits.length()): 0
   Expected (count)            : 1
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(455)]
FAIL!  : tst_QTimeZone::transitionEachZone(America/Cancun@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Asia/Yerevan@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Kaliningrad@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Minsk@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Moscow@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Samara@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Simferopol@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(Europe/Volgograd@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(NET@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(W-SU@2010) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(America/Dawson_Creek@1970) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
FAIL!  : tst_QTimeZone::transitionEachZone(America/Metlakatla@1970) Compared values are not the same
   Actual   (here - stamp / 1000): -3600
   Expected (0)                  : 0
   Loc: [D:/Coding/qt5/qt5/qtbase/tests/auto/corelib/tools/qtimezone/tst_qtimezone.cpp(517)]
```
This needs checking.

Also, this strange `QProcess` failure with detaching really annoys me. I don't understand why I see it only when running all `corelib` tests and not all `corelib/io` tests (there is also quite a bit of them) or not this test alone.

**Comment by dmik** [comment]
Some helpers were not built and some DLL wasn't copied, this explains failures in `QLibrary`, `QSharedMemory` and `QSystemSemaphore` (gone now). The only problematic case is the `QProcess` one. I will try to debug it now.

**Comment by dmik** [comment]
I've created #68 for the `QRegularExpression` failure.
